subprojects/publish/src/main/java/org/gradle/api/publish/internal/versionmapping/DefaultVariantVersionMappingStrategy.java

```diff
@@ -49,13 +52,31 @@ public void fromResolutionOf(String configurationName) {
     }
 
     @Override
-    public String maybeResolveVersion(String group, String module) {
+    public ModuleVersionIdentifier maybeResolveVersion(String group, String module) {
         if (usePublishedVersions && targetConfiguration != null) {
-            Set<? extends ResolvedComponentResult> resolvedComponentResults = targetConfiguration.getIncoming().getResolutionResult().getAllComponents();
+            ResolutionResult resolutionResult = targetConfiguration
+                .getIncoming()
+                .getResolutionResult();
+            Set<? extends ResolvedComponentResult> resolvedComponentResults = resolutionResult.getAllComponents();
             for (ResolvedComponentResult selected : resolvedComponentResults) {
                 ModuleVersionIdentifier moduleVersion = selected.getModuleVersion();
                 if (moduleVersion != null && group.equals(moduleVersion.getGroup()) && module.equals(moduleVersion.getName())) {
-                    return moduleVersion.getVersion();
+                    return moduleVersion;
+                }
+            }
+            // If we reach this point is means we have a dependency which doesn't belong to the resolution result
+            // Which can mean two things:
+            // 1. the graph used to get the resolved version has nothing to do with the dependencies we're trying to get versions for (likely user error)
+            // 2. the graph contains first-level dependencies which have been substituted (likely) so we're going to iterate on dependencies instead
+            Set<? extends DependencyResult> allDependencies = resolutionResult.getAllDependencies();
```

Comment:
Because we look for substitutions only on first level deps, is it not better to request `resolutionResult.getRoot().getDependencies()` ?

Reply:
No, otherwise we would have looked for dependencies in the first iteration already. The thing is that the configuration you resolve is different from the configuration which declares dependencies. It _could_ be totally unrelated. In practice there are not many chances this happens, but it could be that you get a dependency version from a different graph which has this dependency as a non direct one.